cpp/json_ffi/image_utils.cc

```diff
@@ -0,0 +1,157 @@
+  size_t size_;
+  size_t pos_;
+};
+
+size_t Base64DecodedSize(const std::string& base64_str) {
+  size_t len = base64_str.size();
+  size_t padding = 0;
+  if (base64_str[len - 1] == '=') {
+    padding++;
+  }
+  if (base64_str[len - 2] == '=') {
+    padding++;
+  }
+  return 3 * len / 4 - padding;
+}
+
+unsigned char* LoadImageFromBase64(std::string base64_str, int* width, int* height,
+                                   int* num_channels, std::string* err) {
+  MemoryBufferStream stream(base64_str.c_str(), base64_str.size());
+  tvm::support::Base64InStream base64_stream(&stream);
+  size_t decoded_size = Base64DecodedSize(base64_str);
+  unsigned char* decoded = new unsigned char[decoded_size]();
+  base64_stream.InitPosition();
+  base64_stream.Read((void*)decoded, decoded_size);
+  unsigned char* image_data =
+      stbi_load_from_memory(decoded, decoded_size, width, height, num_channels, 3);
+  delete[] decoded;
+  if (!image_data) {
+    *err += stbi_failure_reason();
+  }
+  return image_data;
+}
+
+NDArray ClipPreprocessor(unsigned char* image_data, int width, int height, int target_size,
+                         DLDevice device, std::string* err) {
+  // Resize
+  const int short_side = width < height ? width : height;
+  const int long_side = width > height ? width : height;
+  const int new_short_side = target_size;
+  const int new_long_side = (int)(new_short_side * (long_side / (float)short_side));
+  const int new_width = width < height ? new_short_side : new_long_side;
+  const int new_height = width > height ? new_short_side : new_long_side;
+
+  float* processed_image_data = new float[new_width * new_height * 3]();
+
+  // Bilinear Interpolation
+  for (int y = 0; y < new_height; y++) {
+    for (int x = 0; x < new_width; x++) {
+      const float x_ratio = float(width - 1) / new_width;
+      const float y_ratio = float(height - 1) / new_height;
+      const int x1 = int(x_ratio * x);
+      const int y1 = int(y_ratio * y);
+      const int x2 = x1 + 1;
+      const int y2 = y1 + 1;
+      const float x_diff = x_ratio * x - x1;
+      const float y_diff = y_ratio * y - y1;
+      for (int c = 0; c < 3; c++) {
+        const uint8_t top_left = image_data[(y1 * width + x1) * 3 + c];
+        const uint8_t top_right = image_data[(y1 * width + x2) * 3 + c];
+        const uint8_t bottom_left = image_data[(y2 * width + x1) * 3 + c];
+        const uint8_t bottom_right = image_data[(y2 * width + x2) * 3 + c];
+        processed_image_data[(y * new_width + x) * 3 + c] =
+            (float)(int(top_left * (1 - x_diff) * (1 - y_diff) + top_right * x_diff * (1 - y_diff) +
+                        bottom_left * y_diff * (1 - x_diff) + bottom_right * x_diff * y_diff));
+      }
+    }
+  }
+  stbi_image_free(image_data);
+
+  // Center crop
+  const int crop_x = (new_width - target_size) / 2;
+  const int crop_y = (new_height - target_size) / 2;
+  float* cropped_image_data = new float[target_size * target_size * 3]();
```

Comment:
avoid using new, instead use std::vector<float> cropped_imagte_data(size);

They get automatically released when going out of scope